finesse/hardware/dp9800.py

```diff
@@ -0,0 +1,263 @@
+        Then calculate the BCC and compare it to that transmitted.
+
+        Args:
+            data: the message to check
+
+        Raises:
+            DP9800Error: Malformed message received from device
+        """
+        if data[0] != 2:  # STX
+            raise DP9800Error("Start transmission character not detected")
+        if data[-3] != 3:  # ETX
+            raise DP9800Error("End transmission character not detected")
+        if data[-1] != 0:  # NUL
+            raise DP9800Error("Null terminator not detected")
+
+        bcc = data[-2]
+        bcc_chars = data[1:-2]
+        byte_sum = 0
+        for byte in bcc_chars:
+            byte_sum ^= byte
+
+        if byte_sum != bcc:
+            raise DP9800Error("BCC check failed")
+
+    def parse(self, data: bytes) -> list[Decimal]:
+        """Parse temperature data read from the DP9800.
+
+        The sequence of bytes is translated into a list of ASCII strings
+        representing each of the temperatures, and finally into floats.
+
+        Returns:
+            vals: A list of Decimals containing the temperature values recorded
+                  by the DP9800 device.
+        """
+        if data == b"":
+            return []
+        else:
+            data_ascii = data.decode("ascii")
+            vals_str = [""] * (self.NUM_CHANNELS + 1)
+            vals = [Decimal(0.0)] * (self.NUM_CHANNELS + 1)
+            offset = 3  # offset of temperature values from start of message
+            width = 7  # width of temperature strings
+            for i in range(self.NUM_CHANNELS + 1):
+                vals_str[i] = data_ascii[
+                    self.NUM_CHANNELS * i
+                    + offset : self.NUM_CHANNELS * i
+                    + (offset + width)
+                ]
+                vals[i] = Decimal(vals_str[i])
+
+            sysflag = bin(int(data_ascii[-5:-3], 16))
+            self._sysflag = sysflag[2:]
+
+        return vals[1:]
+
+    def write(self, command: bytes) -> int:
+        r"""Write a message to the DP9800.
+
+        Format:
+
+          EOT T ENQ
+
+        EOT: End of Transmission (ASCII 4)
+        ENQ: Enquiry (ASCII 5)
+
+        Likely to just be used to initiate a read operation, triggered
+        with the command \x04T\x05.
+
+        Args:
+            command: The command to write to the device
+
+        Returns:
+            val: Number of bytes written to the device
+        """
+        num_bytes_written = self.serial.write(command)
+        logging.info(f"Wrote {num_bytes_written} bytes to DP9800")
+        return num_bytes_written
+
+    def send_temperatures(self) -> None:
+        """Perform the complete process of reading from the DP9800.
+
+        Writes to the DP9800 requesting a read operation.
+        Reads the raw data from the DP9800.
+        Parses the data and broadcasts the temperatures.
+        """
+        time_now = datetime.now().timestamp()
```

Comment:
I was thinking that this might be more relevant when writing to the log and/or csv file (which contains the time as well as all the temperatures read by the DP9800), though I've no idea how they use the data and how precise the timing needs to be (maybe they sync it up with something else?).

Either way, I was thinking it makes sense to just calculate it once and broadcast it to all listeners.